File: src/qsiproject.cpp
#include <QDebug>
#include <QFile>
#include <QDir>
#include <QFileInfo>
#include <QFileInfoList>
#include <QRegularExpression>
#include <QRegularExpressionMatchIterator>
#include <QRegularExpressionMatch>
#include <QStringList>
#include <QTextStream>

#include "qsiproject.h"
#include "util.h"

#define QUOTE_REGEX "\"|'|`"

QSIProject::QSIProject(QObject *parent) : QObject(parent) {
	m_projectDir = "";
	m_projectFilePath = "";
	m_projectFileFormat = QSIProject::UnknownProjectType;
	m_buildDir = "";
}

QSIProject::~QSIProject() {

}

bool QSIProject::open(QString path) {
	m_projectDir = path;
	QFileInfo fileInfo(m_projectDir);
	if(!fileInfo.exists()) {
		errorBox("Project file " + m_projectDir + " doesn't exist!");
		return false;
	}

	if(fileInfo.isDir()) {
		QStringList filters({"*.ssproj", "Cellscript.js", "Cellscript.mjs", "Cellscript.cjs", "game.sgm"});
		QFileInfoList infoList = QDir(m_projectDir).entryInfoList(filters,QDir::Files|QDir::NoDotAndDotDot);

		// look for *.ssproj, Cellscript.js, Cellscript.mjs, Cellscript.cjs, and game.sgm, in that order
		for(int f = 0; f < infoList.length(); f++) {
			QFileInfo fi = infoList.at(f);
			m_projectFilePath = fi.filePath();
			QString fileName = fi.fileName();
			QString suffix = fi.suffix();
			if(suffix == "ssproj") {
				m_projectFileFormat = QSIProject::SSProject;
				break;
			} else if(fileName == "Cellscript.js") {
				m_compiler = "Cell";
				m_projectFileFormat = QSIProject::Cellscript_js;
				break;
			} else if(fileName == "Cellscript.mjs") {
				m_compiler = "Cell";
				m_projectFileFormat = QSIProject::Cellscript_mjs;
				break;
			} else if(fileName == "Cellscript.cjs") {
				m_compiler = "Cell";
				m_projectFileFormat = QSIProject::Cellscript_cjs;
				break;
			} else if(fileName == "game.sgm") {
				m_projectFileFormat = QSIProject::SGM;
				break;
			}
		}
	} else {
		m_projectFilePath = fileInfo.filePath();
	}
	if(m_projectFilePath == "") {
		return false;
	}
	QFile* projectFile = new QFile(m_projectFilePath);
	bool success = prepareProjectFile(projectFile);
	projectFile->close();
	delete projectFile;
	return success;
}

bool QSIProject::save() {
	m_projectFilePath = QDir(m_projectDir).absoluteFilePath(QDir(m_projectDir).dirName() + ".ssproj");
	QFile* projectFile = new QFile(m_projectFilePath);
	if(!projectFile->open(QFile::WriteOnly)) {
		errorBox("Unable to open '" + m_projectFilePath + "': " + projectFile->errorString());
		delete projectFile;
		return false;
	}
	QString ssText;
	QTextStream(&ssText) <<
		"[.ssproj]\r\n" <<
		"author=" << m_author << "\r\n" <<
		"compiler=" << getCompiler() << "\r\n" <<
		"description=" << m_summary << "\r\n" <<
		"mainScript=" << m_script << "\r\n" <<
		"name=" << getName() << "\r\n" <<
		"screenHeight=" << m_height << "\r\n" <<
		"screenWidth=" << m_width << "\r\n";


	bool success = false;
	if(projectFile->write(ssText.toLatin1()) > -1) {
		m_projectFileFormat = QSIProject::SSProject;
		success = true;
	}
	projectFile->close();
	return success;
}

QString QSIProject::getName() {
	if(m_name == "") return QFileInfo(m_projectDir).baseName();
	return m_name;
}

void QSIProject::setName(QString name) {
	m_name = name;
}

QString QSIProject::getResolutionString() {
	QString resolution = "";
	return resolution.sprintf("%dx%d", m_width, m_height);
}

QString QSIProject::getAuthor() {
	return m_author;
}

void QSIProject::setAuthor(QString author) {
	m_author = author;
}

int QSIProject::getWidth() {
	return m_width;
}

void QSIProject::setWidth(int width) {
	m_width = width;
}

int QSIProject::getHeight() {
	return m_height;
}

void QSIProject::setHeight(int height) {
	m_height = height;
}

void QSIProject::setSize(int width, int height) {
	m_width = width;
	m_height = height;
}

int QSIProject::getAPILevel() {
	return m_apiLevel;
}

void QSIProject::setAPILevel(int level) {
	m_apiLevel = level;
}

int QSIProject::getVersion() {
	return m_version;
}

void QSIProject::setVersion(int version) {
	m_version = version;
}

QString QSIProject::getSaveID() {
	return m_saveID;
}

void QSIProject::setSaveID(QString id) {
	m_saveID = id;
}

QString QSIProject::getSummary() {
	return m_summary;
}

void QSIProject::setSummary(QString summary) {
	m_summary = summary;
}

QString QSIProject::getBuildDir() {
	if(m_compiler == "Vanilla") return getPath(false);
	return m_buildDir;
}

void QSIProject::setBuildDir(QString dir) {
	m_buildDir = dir;
}

QString QSIProject::getMainScript() {
	return m_script;
}

void QSIProject::setMainScript(QString path) {
	m_script = path;
}

QString QSIProject::getPath(bool projectFile) {
	if(projectFile) return m_projectFilePath;
	return m_projectDir;
}

void QSIProject::setPath(QString path, bool projectFile) {
	if(projectFile) m_projectFilePath = path;
	else m_projectDir = path;
}

QIcon QSIProject::getIcon() {
	QFileInfo icon_fi = QFileInfo(QDir(m_projectDir), "icon.png");
	if(icon_fi.exists()) {
		return QIcon(icon_fi.canonicalFilePath());
	}
	return QIcon(":/icons/sphere-icon.png");
}

QSIProject::ProjectFileFormat QSIProject::getProjectFormat() {
	return m_projectFileFormat;
}

QString QSIProject::getCompiler() {
	return m_compiler;
}

void QSIProject::setCompiler(QString compiler) {
	m_compiler = compiler;
}

bool QSIProject::prepareProjectFile(QFile* projectFile) {
	if(projectFile == nullptr) return false;
	if(!projectFile->isOpen()) {
		if(!projectFile->open(QFile::ReadOnly)) {
			errorBox("Unable to open project file in '" + m_projectFilePath + "'");
			return false;
		}
	}
	switch(m_projectFileFormat) {
	case QSIProject::SSProject:
		return readSSProj(projectFile);
	case QSIProject::Cellscript_js:
	case QSIProject::Cellscript_mjs:
	case QSIProject::Cellscript_cjs:
		return readCellscript(projectFile);
	case QSIProject::SGM:
		return readSGM(projectFile);
	default:
		// no project file in directory
		return false;
	}
}

bool QSIProject::readSSProj(QFile* projectFile) {
	QTextStream stream(projectFile);
	while(!stream.atEnd()) {
		QStringList arr = stream.readLine().split("=");
		if(arr.length() != 2) continue;
		QString key = arr.at(0);
		QString value = arr.at(1);

		if(key == "author") m_author = value;
		else if(key == "buildDir") m_buildDir = QDir(m_projectDir).absoluteFilePath(value);
		else if(key == "compiler") {
			m_compiler = value;
		}
		else if(key == "description") m_summary = value;
		else if(key == "mainScript") m_script = value;
		else if(key == "name") m_name = value;
		else if(key == "screenHeight") m_height = value.toInt();
		else if(key == "screenWidth") m_width = value.toInt();
	}
	return true;
}

QString QSIProject::getCellscriptStringValue(QString cellscriptStr, QString key, QString defaultValue) {
	QString value = defaultValue;
	QString reStr = "(" + key + "):\\s*(" + QUOTE_REGEX + ")(.*)(" + QUOTE_REGEX + "),";
	QRegularExpression re(reStr, QRegularExpression::OptimizeOnFirstUsageOption);
	QStringList captureList = re.match(cellscriptStr).capturedTexts();
	if(captureList.length() > 4)
		value = captureList.at(3);

	return value
		.replace("\\'", "'")
		.replace("\\\"", "\"")
		.replace("\\`","`");
}

int QSIProject::getCellscriptIntValue(QString cellscriptStr, QString key, int defaultValue) {
	int value = defaultValue;
	QString reStr = "(" + key + "):\\s*(\\d+).*,";
	QRegularExpression re(reStr, QRegularExpression::OptimizeOnFirstUsageOption);
	QStringList captureList = re.match(cellscriptStr).capturedTexts();
	if(captureList.length() >= 3) {
		bool ok = false;
		value = captureList.at(2).toInt(&ok);
		if(!ok) value = defaultValue;
	}
	return value;
}

bool QSIProject::readCellscript(QFile* projectFile) {
	QTextStream stream(projectFile);
	QString text = stream.readAll();
	m_name = getCellscriptStringValue(text, "name");
	m_author = getCellscriptStringValue(text, "author");
	m_version = getCellscriptIntValue(text, "version", 1);
	m_apiLevel = getCellscriptIntValue(text, "apiLevel", 1);
	m_saveID = getCellscriptStringValue(text, "saveID");
	m_summary = getCellscriptStringValue(text, "summary");
	m_script = getCellscriptStringValue(text, "main");
	QStringList resolution = getCellscriptStringValue(text, "resolution").split("x");
	bool ok = false;
	if(resolution.length() != 2) {
		m_width = -1;
		m_height = -1;
		return true;
	}
	m_width = resolution.at(0).toInt(&ok);
	if(!ok) m_width = -1;
	m_height = resolution.at(1).toInt(&ok);
	if(!ok) m_height = -1;

	if(m_width < 0 || m_height < 0) {
		qDebug() << m_name << "has an invalid resolution:" << resolution.at(0) << "x" << resolution.at(1);
	}
	m_compiler = "Cell";

	/*qDebug().nospace().noquote() <<
		"Project: " << m_name << "\n" <<
		"Author: " << m_author << "\n" <<
		"Version: " << m_version << "\n" <<
		"API Level: " << m_apiLevel << "\n" <<
		"Save ID: " << m_saveID << "\n" <<
		"Summary: " << m_summary << "\n" <<
		"Resolution: " << m_width << "x" << m_height << "\n" <<
		"Main script: " << m_script << "\n";*/
	m_buildDir = QFileInfo(*projectFile).dir().absoluteFilePath("dist/");
	return true;
}

bool QSIProject::readSGM(QFile *projectFile) {
	QTextStream stream(projectFile);
	while(!stream.atEnd()) {
		QString line = stream.readLine();
		if(line == "") continue;
		QString key = line.section("=", 0, 0);
		QString value = line.section("=", 1);
		if(key == "name") m_name = value;
		else if(key == "author") m_author = value;
		else if(key == "description") m_summary = value;
		else if(key == "screen_width") m_width = value.toInt();
		else if(key == "screen_height") m_height = value.toInt();
		else if(key == "script") m_script = value;
	}
	m_compiler = "Vanilla";
	m_buildDir = QFileInfo(*projectFile).dir().path();
	return true;
}

File: qsiproject.py
from enum import Enum
from io import TextIOWrapper
import re

from PySide6.QtGui import QIcon
from PySide6.QtCore import QDir, QFileInfo
from PySide6.QtWidgets import QMessageBox

_file_filters = ("*.ssproj", "Cellscript.js", "Cellscript.mjs", "Cellscript.cjs", "game.sgm")

class ProjectType(Enum):
	Unknown = 0
	SSProject = 1
	Cellscript_js = 2
	Cellscript_mjs = 3
	Cellscript_cjs = 4
	SGM = 5

class QSIProject:
	name: str
	author: str
	width: int
	height: int
	apiLevel: int
	version: int
	projectType: ProjectType
	projectFilePath: str
	saveID: str
	summary: str
	buildDir: str
	script: str
	projectDir: str
	compiler: str

	def __init__(self):
		self.projectType = ProjectType.Unknown
		self.projectFilePath = None
		self.name = None
		self.author = None
		self.width = None
		self.height = None
		self.buildDir = None


	def open(self, path: str) -> bool:
		self.projectDir = path
		fileInfo = QFileInfo(path)
		if not fileInfo.exists():
			QMessageBox.critical(None, "Error!", "Project path %s does not exist" % path)
			return False
		
		if fileInfo.isDir():
			infoList = QDir(path).entryInfoList(_file_filters, QDir.Filter.Files|QDir.Filter.NoDotAndDotDot)
			for info in infoList:
				self.projectFilePath = info.filePath()
				filename = info.fileName()
				if info.suffix() == "ssproj":
					self.projectType = ProjectType.SSProject
				elif filename == "Cellscript.js":
					self.projectType = ProjectType.Cellscript_js
					self.compiler = "Cell"
				elif filename == "Cellscript.cjs":
					self.projectType = ProjectType.Cellscript_cjs
					self.compiler = "Cell"
				elif filename == "Cellscript.mjs":
					self.projectType = ProjectType.Cellscript_cjs
					self.compiler = "Cell"
				elif filename == "game.sgm":
					self.projectType = ProjectType.SGM
		else:
			self.projectFilePath = fileInfo.filePath()
			self.projectDir = fileInfo.dir().path()
		
		if self.projectFilePath == "" or self.projectFilePath is None:
			print(f"Unable to get project info for path {path} - skipping")
			return False

		self.buildDir = self.projectDir
		with open(self.projectFilePath, "r") as projectFile:
			return self._prepareProjectFile(projectFile)


	def getIcon(self) -> QIcon:
		iconInfo = QFileInfo(QDir(self.projectDir), "icon.png");
		if iconInfo.exists():
			return QIcon(iconInfo.canonicalFilePath())
		return QIcon(":/icons/res/neosphere.png")


	def getResolutionString(self) -> str:
		return "%dx%d" % (self.width, self.height)


	def _getCellscriptStringValue(self, cellscriptStr:str, key:str, defaultValue:str = "") -> str:
		matches = re.match(r".*" + key + r"\s*:\s*(['\"`]).*", cellscriptStr, re.DOTALL|re.MULTILINE)
		if matches is None:
			return defaultValue
		quoteChar = matches.group(1)[0]
		startIndex = matches.start(1)
		scriptClipped = cellscriptStr[startIndex+1:]
		endIndex = -1
		for i in range(len(scriptClipped)):
			if scriptClipped[i] == quoteChar and scriptClipped[i-1] != '\\':
				endIndex = i
				break
		return scriptClipped[:endIndex]

	def _getCellscriptIntValue(self, cellscriptStr:str, key:str, defaultValue:int = 0) -> int:
		matches = re.findall(r".*(" + key + r")\s*:\s*(\d+).*", cellscriptStr, re.DOTALL|re.MULTILINE)
		if len(matches) == 0:
			return defaultValue
		value:str = matches[0][1]
		return int(value) if value.isdigit() else defaultValue

	def _prepareProjectFile(self, projectFile:TextIOWrapper) -> bool:
		match self.projectType:
			case ProjectType.SSProject:
				return self._readSSProj(projectFile)
			case ProjectType.Cellscript_js | ProjectType.Cellscript_mjs | ProjectType.Cellscript_cjs:
				return self._readCellscript(projectFile)
			case ProjectType.SGM:
				return self._readSGM(projectFile)


	def _readSSProj(self, projectFile:TextIOWrapper) -> bool:
		lines = projectFile.readlines()
		for line in lines:
			if line == "" or line.find("=") == -1:
				continue
			parts = line.split("=")
			match parts[0]:
				case "author":
					self.author = parts[1]
				case "buildDir":
					self.buildDir = QDir(self.projectDir).absoluteFilePath(parts[1])
				case "compiler":
					self.compiler = parts[1]
				case "description":
					self.summary = parts[1]
				case "mainScript":
					self.script = parts[1]
				case "name":
					self.name = parts[1]
				case "screenHeight":
					self.height = int(parts[1])
				case "screenWidth":
					self.width = int(parts[1])
		return True


	def _readCellscript(self, projectFile:TextIOWrapper) -> bool:
		scriptStr = projectFile.read()
		self.name = self._getCellscriptStringValue(scriptStr, "name")
		self.author = self._getCellscriptStringValue(scriptStr, "author")
		self.version = self._getCellscriptIntValue(scriptStr, "version", 1)
		self.apiLevel = self._getCellscriptIntValue(scriptStr, "apiLevel", 1)
		self.saveID = self._getCellscriptStringValue(scriptStr, "saveID")
		self.summary = self._getCellscriptStringValue(scriptStr, "summary")
		self.script = self._getCellscriptStringValue(scriptStr, "main")
		resolution = self._getCellscriptStringValue(scriptStr, "resolution")
		resolutionArr = resolution.split("x")

		if len(resolutionArr) != 2:
			# expects format <width>x<height>, missing x or more than one x
			QMessageBox.critical(None, "Error", f"Project {self.name} has an invalid resolution string: {resolution}")
			return False
		
		self.width = int(resolutionArr[0]) if resolutionArr[0].isdigit() else -1
		self.height = int(resolutionArr[1]) if resolutionArr[1].isdigit() else -1
		
		if self.width < 1 or self.height < 1:
			# width or height are an invalid number
			QMessageBox.critical(None, "Error", f"Project {self.name} has an invalid resolution string: {resolution}")
			return False

		self.compiler = "Cell"
		self.buildDir = QFileInfo(self.projectFilePath).dir().absoluteFilePath("dist/")
		return True


	def _readSGM(self, projectFile:TextIOWrapper) -> bool:
		lines = projectFile.readlines()
		for line in lines:
			if line == "":
				continue
			parts = line.split("=", 1)
			match parts[0]:
				case "name":
					self.name = parts[1]
				case "author":
					self.author = parts[1]
				case "description":
					self.summary = parts[1]
				case "screen_width":
					self.width = int(parts[1])
				case "screen_height":
					self.height = int(parts[1])
				case "script":
					self.script = parts[1]
		return True

	def __repr__(self) -> str:
		return repr((self.name, self.author))
